arch/arm/core/elf.c

```diff
@@ -11,79 +12,333 @@
+	uint32_t upper, lower;
 
-	addend = upper & ARM_BL_BLX_UPPER_S_BIT ? addend >> 1 : addend >> 2;
+	upper = MEM2THM16OPCODE(*(uint16_t *)loc);
+	lower = MEM2THM16OPCODE(*(uint16_t *)(loc + 2));
 
-	upper &= ARM_BL_BLX_HDR_MASK;
-	lower &= ARM_BL_BLX_HDR_MASK;
-	upper |= (addend >> ARM_BL_BLX_ADDEND_SIZE) & ARM_BL_BLX_ADDEND_MASK;
-	lower |= addend & ARM_BL_BLX_ADDEND_MASK;
+	/* MOVT/MOVW instructions encoding in Thumb-2 */
+	*offset = ((upper & MASK_THM_MOV_IMM4) << SHIFT_THM_MOV_IMM4) |
+		((upper & MASK_THM_MOV_I) << SHIFT_THM_MOV_I) |
+		((lower & MASK_THM_MOV_IMM3) >> SHIFT_THM_MOV_IMM3) | (lower & MASK_THM_MOV_IMM8);
+	*offset = sign_extend(*offset, SHIFT_THM_MOVS_SIGN);
+	*offset += sym_base_addr;
+
+	if (reloc_type == R_ARM_THM_MOVT_PREL || reloc_type == R_ARM_THM_MOVW_PREL_NC) {
+		*offset -= loc;
+	}
+	if (reloc_type == R_ARM_THM_MOVT_ABS || reloc_type == R_ARM_THM_MOVT_PREL) {
+		*offset >>= 16;
+	}
 
-	*((uint16_t *)opaddr) = upper;
-	*(((uint16_t *)opaddr) + 1) = lower;
+	upper = (uint16_t)((upper & (MASK_THM_MOV_11110|MASK_THM_MOV_100100)) |
+			((*offset & (MASK_THM_MOV_IMM4<<SHIFT_THM_MOV_IMM4)) >> SHIFT_THM_MOV_IMM4) |
+			((*offset & (MASK_THM_MOV_I<<SHIFT_THM_MOV_I)) >> SHIFT_THM_MOV_I));
+	lower = (uint16_t)((lower & (MASK_THM_MOV_0|MASK_THM_MOV_RD)) |
+			((*offset & (MASK_THM_MOV_IMM3>>SHIFT_THM_MOV_IMM3)) << SHIFT_THM_MOV_IMM3) |
+			(*offset & MASK_THM_MOV_IMM8));
+	*(uint16_t *)loc = OPCODE2THM16MEM(upper);
+	*(uint16_t *)(loc + 2) = OPCODE2THM16MEM(lower);
 }
 
 /**
- * @brief Architecture specific function for relocating partially linked (static) elf
+ * Decoder functions prefixed by '_thm_' means that they are Thumb instructions specific.
+ * Do NOT mix them with not 'Thumb instructions' in the nelow switch/case: they are not
+ * intended to work together.
  *
- * Elf files contain a series of relocations described in a section. These relocation
- * instructions are architecture specific and each architecture supporting extensions
- * must implement this.
+ * loc = dstsec->sh_addr + rel->r_offset;
  *
- * The relocation codes for arm are well documented
- * https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst#relocation
+ * loc : addr of opcode
+ * sym_base_addr : addr of symbol if undef else addr of symbol section + sym value
+ * sym_name: symbol name
  */
-void arch_elf_relocate(elf_rela_t *rel, uintptr_t opaddr, uintptr_t opval)
+static int apply_relocate(elf_word reloc_type, uint32_t loc, uint32_t sym_base_addr,
+			const char *sym_name, uintptr_t load_bias)
 {
-	elf_word reloc_type = ELF32_R_TYPE(rel->r_info);
+	int ret = 0;
+	int32_t offset;
+
+	LOG_DBG("%s: %d %x %x %s", __func__, reloc_type, loc, sym_base_addr, sym_name);
 
 	switch (reloc_type) {
+	case R_ARM_NONE:
+		/* no-op */
+		break;
+
 	case R_ARM_ABS32:
-		/* Add the addend stored at opaddr to opval */
-		opval += *((uint32_t *)opaddr);
+		/* fall-through */
+	case R_ARM_TARGET1:
+		*(uint32_t *)loc += sym_base_addr;
```

Comment:
Here is an answer from my technical expert (what he meant is that is already done, but in several stages):


As in '_(S + A) | T_',  the '_| T_' does mean that we must set the bit0 if we are in Thumb & a function.
This is done by LLEXT while we are computing '_link_addr_' in '_llext_link_' which becomes '_sym_addr_' in '_arch_elf_relocate()_'

`else if (ELF_ST_TYPE(sym.st_info) == STT_SECTION || `
`                                      ELF_ST_TYPE(sym.st_info) == STT_FUNC || `
`                                      ELF_ST_TYPE(sym.st_info) == STT_OBJECT) { `
`                                      /* Link address is relative to the start of the section */ `
`                                      link_addr = (uintptr_t)ext->mem[ldr->sect_map[sym.st_shndx]] `
`                                                 + sym.st_value;`

The internal extension symbols are handled by the `+ sym.st_value;`
The external ones:

`else if (sym.st_shndx == SHN_UNDEF) {`
`                                       /* If symbol is undefined, then we need to look it up */`
`                                       link_addr = (uintptr_t)llext_find_sym(NULL, name);`

'_llext_find_sym()_' returns the correct address because it has been register via the EXPORT_SYNBOL macro which actually handles the bit0 in Thumb.
Here is an example with the movwmovt.llext test:
![test_start](https://github.com/zephyrproject-rtos/zephyr/assets/162981852/ddd3a290-6190-4818-8caf-2d5908161425)
![test_end](https://github.com/zephyrproject-rtos/zephyr/assets/162981852/91b13332-c635-4101-b4a2-c8f3496764f4)
For the FUNC type '_test_func_' symbol, '_sym.st_value_' equals 1 but 0 for the others. Same for '_test_entry_' where '_sym.st_value_' is odd.